00-programming-fundamentals/08-intro-to-automated-tests.md

```diff
@@ -60,24 +66,273 @@ true
+  def test_creation_of_die
+    @die = Die.new
+    # The class of @die should be Die
+    assert @die.class == Die, "There must be a Die class."
+  end
+end
+```
+
+Notice the line `assert @die.class, Die`.  The `assert` method takes two required arguments and an optional message.  If the two required arguments are equal, the test passes, otherwise it fails.  
+
+When we run this with `ruby die_test.rb` we get:
+
+```bash
+ruby die_test.rb
+MiniTest::Unit::TestCase is now Minitest::Test. From die_test.rb:6:in `<main>'
+Run options: --seed 29986
+
+# Running:
+
+E
+
+Finished in 0.001219s, 820.3446 runs/s, 0.0000 assertions/s.
+
+  1) Error:
+TestDie#test_creation_of_die:
+NameError: uninitialized constant TestDie::Die
+Did you mean?  Dir
+    die_test.rb:8:in `test_creation_of_die'
+
+1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
+```
+
+**This is a good thing.**  We have our first **red** test.  There's an error because we haven't created the Die class yet.
+
+Lets do so:
+
+```ruby
+# die.rb
+
+class Die
+
+end
+```
+
+Now `ruby die_test.rb` gets us:
+
+```bash
+MiniTest::Unit::TestCase is now Minitest::Test. From die_test.rb:6:in `<main>'
+Run options: --seed 61965
+
+# Running:
+
+.
+
+Finished in 0.001211s, 825.7638 runs/s, 825.7638 assertions/s.
+
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+```
+
+Now we have our first **green/passing** test.  
+
+There are a [number of assertions](https://gist.github.com/rastasheep/4248006#Minitest::Unit::TestCase) in Minitest beyond the assert method.  
+
+#### Practice Exercise
+
+Now we need to test that we can roll the die.  Write a test that checks the roll method of the Die class.  When a die is rolled it should return a number between 1 and 6 inclusive.  You can use the `assert_operator` method.  
+
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here.](https://github.com/AdaGold/die/blob/solution/tests/die_test.rb)
+
+### Spec Style Testing
+
+Spec style tests look very different, but read in a more English friendly fashion.  We'll create the test with a file named `die_spec.rb`.
+
+```ruby
+# die_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+```
+
+Up till now this all looks the same, but next it changes a bit.  Instead of creating a class we'll write a `describe` block which identifies a group of tests that belong together, usually because they are testing the same thing.  Inside the `describe` block are `it` blocks.  Each `it` block is a test.   
+
+```ruby
+# die_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+
+describe "Testing Die Class" do
+  it "You can create an instance of Die" do
+    @die = Die.new
+
+    @die.class.must_equal Die
+  end
+end
+
+```
+
+Notice the line `@die.class.must_equal Die`.  Look at how english-friendly it is.  The idea with **BDD** is the that with our tests we are defining specifications or behaviors our code should follow.  
```

Comment:
For the second sentence, maybe something more like "Observe how this reads more easily from left to right than our original assert test."